File: lib/Repository/HttpRepository.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../Utils/AppSession.dart';
import '../Utils/Utils.dart';
import 'BaseRepository.dart';

/// HTTP wrapper chung cho các Repository.
class HttpRepository extends BaseRepository {
  static const Duration _timeout = Duration(seconds: 15);

  Map<String, String> _buildHeaders({Map<String, String>? extra}) {
    final headers = <String, String>{
      'Content-Type': 'application/json',
      ...?extra,
    };
    final token = AppSession.accessToken;
    if (token != null && token.isNotEmpty) {
      headers['Authorization'] = 'Bearer $token';
    }
    return headers;
  }

  Uri _buildUri(String path, [Map<String, dynamic>? query]) {
    final base = Utils.apiUrl;
    final raw = base.endsWith('/') ? base.substring(0, base.length - 1) : base;
    final p = path.startsWith('/') ? path : '/$path';
    final uri = Uri.parse('$raw$p');
    if (query == null || query.isEmpty) return uri;

    final qp = <String, String>{};
    query.forEach((key, value) {
      if (value == null) return;
      final s = value.toString();
      if (s.isEmpty) return;
      qp[key] = s;
    });
    return uri.replace(queryParameters: qp);
  }

  Future<Map<String, dynamic>> getJson(
    String path, {
    Map<String, dynamic>? query,
    Map<String, String>? headers,
  }) async {
    final response = await _safeRequest(
      () => http.get(
        _buildUri(path, query),
        headers: _buildHeaders(extra: headers),
      ),
    );
    return _decodeAsMap(response);
  }

  Future<Map<String, dynamic>> postJson(
    String path, {
    Object? body,
    Map<String, dynamic>? query,
    Map<String, String>? headers,
  }) async {
    final response = await _safeRequest(
      () => http.post(
        _buildUri(path, query),
        headers: _buildHeaders(extra: headers),
        body: body == null ? null : jsonEncode(body),
      ),
    );
    return _decodeAsMap(response);
  }

  Future<Map<String, dynamic>> putJson(
    String path, {
    Object? body,
    Map<String, dynamic>? query,
    Map<String, String>? headers,
  }) async {
    final response = await _safeRequest(
      () => http.put(
        _buildUri(path, query),
        headers: _buildHeaders(extra: headers),
        body: body == null ? null : jsonEncode(body),
      ),
    );
    return _decodeAsMap(response);
  }

  Future<Map<String, dynamic>> deleteJson(
    String path, {
    Object? body,
    Map<String, dynamic>? query,
    Map<String, String>? headers,
  }) async {
    final response = await _safeRequest(
      () => http.delete(
        _buildUri(path, query),
        headers: _buildHeaders(extra: headers),
        body: body == null ? null : jsonEncode(body),
      ),
    );
    return _decodeAsMap(response);
  }

  Future<http.Response> _safeRequest(
    Future<http.Response> Function() requestFn,
  ) async {
    try {
      return await requestFn().timeout(_timeout);
    } on TimeoutException {
      throw Exception('Kết nối hết thời gian chờ, vui lòng thử lại.');
    } on SocketException {
      throw Exception('Không thể kết nối tới máy chủ, vui lòng kiểm tra mạng.');
    } on http.ClientException catch (e) {
      throw Exception('Lỗi mạng: ${e.message}');
    }
  }

  Map<String, dynamic> _decodeAsMap(http.Response response) {
    // Sign out + throw nếu 401.
    super.handleStatus(response.statusCode);

    String _extractMessage(Map<String, dynamic>? payload) {
      if (payload == null) return '';
      final message = payload['message'];
      final error = payload['error'];

      String stringify(dynamic value) {
        if (value is String) return value;
        if (value is List) {
          return value.map(stringify).where((e) => e.isNotEmpty).join(' ; ');
        }
        if (value is Map) {
          return value.values.map(stringify).where((e) => e.isNotEmpty).join(' ; ');
        }
        return value?.toString() ?? '';
      }

      final msg = stringify(message);
      if (msg.isNotEmpty) return msg;
      final err = stringify(error);
      if (err.isNotEmpty) return err;
      return '';
    }

    Map<String, dynamic>? decoded;
    if (response.body.isNotEmpty) {
      try {
        final payload = jsonDecode(response.body);
        if (payload is Map<String, dynamic>) {
          decoded = payload;
        } else {
          decoded = <String, dynamic>{
            'statusCode': response.statusCode,
            'message': '',
            'data': payload,
          };
        }
      } catch (_) {
        throw Exception('Phản hồi không hợp lệ từ máy chủ.');
      }
    }

    // Kiểm tra HTTP code không thành công.
    if (response.statusCode < 200 || response.statusCode >= 300) {
      final msg = _extractMessage(decoded);
      throw Exception(
        msg.isNotEmpty ? msg : 'Yêu cầu thất bại (HTTP ${response.statusCode}).',
      );
    }

    // Một số API gói statusCode trong payload, kiểm tra thêm.
    final innerStatus = (decoded?['statusCode'] as num?)?.toInt();
    if (innerStatus != null && (innerStatus < 200 || innerStatus >= 300)) {
      final msg = _extractMessage(decoded);
      throw Exception(
        msg.isNotEmpty ? msg : 'Yêu cầu thất bại (mã $innerStatus).',
      );
    }

    if (decoded != null) {
      return decoded!;
    }

    // No-content responses.
    return <String, dynamic>{
      'statusCode': response.statusCode,
      'message': '',
      'data': null,
    };
  }
}

File: mobile_restaurant/repositories/HttpRepository.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import axios, { type AxiosInstance, type InternalAxiosRequestConfig } from 'axios';
import { apiUrl } from '../utils/Utils';

const SESSION_ACCESS_KEY = 'session_access';
const SESSION_REFRESH_KEY = 'session_refresh';

/**
 * HTTP wrapper using Axios.
 * Automatically handles auth tokens, token refresh, timeouts, and standard backend error parsing.
 */

// Create Axios Instance
export const apiClient: AxiosInstance = axios.create({
  baseURL: apiUrl.endsWith('/') ? apiUrl.slice(0, -1) : apiUrl,
  timeout: 15000,
  headers: {
    'Content-Type': 'application/json',
  },
});

function isTokenExpired(token: string): boolean {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return true;
    
    const base64Url = parts[1];
    const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
    
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    let buffer = '';
    const str = base64.replace(/=+$/, '');
    for (let i = 0, bc = 0, bs = 0; i < str.length; i++) {
      const char = str.charAt(i);
      const idx = chars.indexOf(char);
      if (idx === -1) continue;
      bs = bc % 4 ? bs * 64 + idx : idx;
      if (bc++ % 4) {
        buffer += String.fromCharCode(255 & (bs >> ((-2 * bc) & 6)));
      }
    }
    
    const payload = JSON.parse(buffer);
    if (payload && typeof payload.exp === 'number') {
      const expiry = payload.exp * 1000;
      // Consider expired 10 seconds before actual expiry
      return Date.now() >= (expiry - 10000);
    }
    return false;
  } catch (e) {
    return true;
  }
}

// Mutex to prevent multiple concurrent refresh calls
let isRefreshing = false;
let refreshSubscribers: ((token: string) => void)[] = [];

function onTokenRefreshed(newToken: string) {
  refreshSubscribers.forEach((cb) => cb(newToken));
  refreshSubscribers = [];
}

function addRefreshSubscriber(cb: (token: string) => void) {
  refreshSubscribers.push(cb);
}

async function tryRefreshToken(): Promise<string | null> {
  const refreshToken = await AsyncStorage.getItem(SESSION_REFRESH_KEY);
  if (!refreshToken) return null;

  try {
    // Call the refresh endpoint directly with axios (not apiClient) to avoid interceptor loops
    const baseURL = apiUrl.endsWith('/') ? apiUrl.slice(0, -1) : apiUrl;
    const response = await axios.post(`${baseURL}/auth/refresh`, { refreshToken }, {
      timeout: 10000,
      headers: { 'Content-Type': 'application/json' },
    });

    const data = response.data?.data ?? response.data;
    const newAccessToken = data?.accessToken;

    if (newAccessToken) {
      await AsyncStorage.setItem(SESSION_ACCESS_KEY, newAccessToken);
      return newAccessToken;
    }
    return null;
  } catch (err) {
    console.warn('Token refresh failed:', err);
    // Refresh failed — clear all session data
    await AsyncStorage.multiRemove([SESSION_ACCESS_KEY, SESSION_REFRESH_KEY, 'session_user']).catch(() => {});
    return null;
  }
}

// Request Interceptor: Attach Token (with auto-refresh)
apiClient.interceptors.request.use(
  async (config) => {
    let token = await AsyncStorage.getItem(SESSION_ACCESS_KEY);

    if (token && isTokenExpired(token)) {
      // Token is expired or about to expire — try to refresh
      if (!isRefreshing) {
        isRefreshing = true;
        const newToken = await tryRefreshToken();
        isRefreshing = false;

        if (newToken) {
          token = newToken;
          onTokenRefreshed(newToken);
        } else {
          // Refresh failed, clear everything
          token = null;
          refreshSubscribers = [];
        }
      } else {
        // Another request is already refreshing — wait for it
        token = await new Promise<string>((resolve) => {
          addRefreshSubscriber((newToken: string) => resolve(newToken));
        });
      }
    }

    if (token && config.headers) {
      config.headers.Authorization = `Bearer ${token}`;
    }

    return config;
  },
  (error) => Promise.reject(error)
);

function extractMessage(payload: any): string {
  if (!payload) return '';
  const stringify = (value: any): string => {
    if (typeof value === 'string') return value;
    if (Array.isArray(value)) return value.filter(Boolean).join(' ; ');
    if (value && typeof value === 'object') return Object.values(value).filter(Boolean).join(' ; ');
    return value?.toString() ?? '';
  };
  const msg = stringify(payload.message);
  if (msg && msg !== 'null') return msg;
  const err = stringify(payload.error);
  if (err && err !== 'null') return err;
  return '';
}

// Response Interceptor: Error Handling (with 401 retry after refresh)
apiClient.interceptors.response.use(
  (response) => {
    // Backend success format { statusCode, message, data }
    return response.data ?? { statusCode: response.status, message: '', data: null };
  },
  async (error) => {
    if (error.code === 'ECONNABORTED') {
      return Promise.reject(new Error('Kết nối hết thời gian chờ, vui lòng thử lại.'));
    }
    if (!error.response) {
      return Promise.reject(new Error('Không thể kết nối tới máy chủ, vui lòng kiểm tra mạng.'));
    }

    const originalRequest = error.config;
    const { status, data } = error.response;
    const msg = extractMessage(data);

    // If 401 and we haven't retried yet, try refreshing the token once
    if (status === 401 && !originalRequest._hasRetried) {
      originalRequest._hasRetried = true;

      const newToken = await tryRefreshToken();
      if (newToken) {
        originalRequest.headers.Authorization = `Bearer ${newToken}`;
        return apiClient(originalRequest);
      }
    }
    
    if (status === 401) {
      return Promise.reject(new Error(msg || 'Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại.'));
    }

    return Promise.reject(new Error(msg || `Yêu cầu thất bại (HTTP ${status}).`));
  }
);

// ===== Public API =====

export async function getJson(
  path: string,
  options?: { query?: Record<string, any>; headers?: Record<string, string> },
): Promise<any> {
  const p = path.startsWith('/') ? path : `/${path}`;
  return apiClient.get(p, { params: options?.query, headers: options?.headers });
}

export async function postJson(
  path: string,
  options?: { body?: any; query?: Record<string, any>; headers?: Record<string, string> },
): Promise<any> {
  const p = path.startsWith('/') ? path : `/${path}`;
  return apiClient.post(p, options?.body, { params: options?.query, headers: options?.headers });
}

export async function putJson(
  path: string,
  options?: { body?: any; query?: Record<string, any>; headers?: Record<string, string> },
): Promise<any> {
  const p = path.startsWith('/') ? path : `/${path}`;
  return apiClient.put(p, options?.body, { params: options?.query, headers: options?.headers });
}

export async function deleteJson(
  path: string,
  options?: { body?: any; query?: Record<string, any>; headers?: Record<string, string> },
): Promise<any> {
  const p = path.startsWith('/') ? path : `/${path}`;
  return apiClient.delete(p, { data: options?.body, params: options?.query, headers: options?.headers });
}
